ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,507 @@
+from lxml import etree
+from resumable import DownloadCheck, DownloadError, urlretrieve
+import yaml
+
+from .systemd import Manager as SystemManager, NoSuchUnit
+
+
+class InvalidFile(Exception):
+    pass
+
+
+class InvalidPackageMetadata(Exception):
+    pass
+
+
+class InvalidPackageType(Exception):
+    pass
+
+
+class NoSuchPackage(Exception):
+    pass
+
+
+class InvalidHandlerType(Exception):
+    pass
+
+
+class Remote:
+    def __init__(self, id, name, url):
+        self.id = id
+        self.name = name
+        self.url = url
+
+    @classmethod
+    def from_file(cls, path):
+        with open(path, 'r') as f:
+            d = yaml.safe_load(f.read())
+
+            try:
+                return cls(d['id'], d['name'], d['url'])
+
+            except KeyError as e:
+                raise InvalidFile(
+                    'Remote file is missing a {} key: {}'.format(e, path))
+
+    def to_file(self, path):
+        with open(path, 'w') as f:
+            d = {'id': self.id, 'name': self.name, 'url': self.url}
+            f.write(yaml.safe_dump(d, default_flow_style=False))
+
+
+class MetaRegistry(type):
+    def __new__(mcs, name, bases, attrs, **kwargs):
+        cls = super().__new__(mcs, name, bases, attrs)
+
+        try:
+            baseclass = bases[0]
+            baseclass.registered_types[cls.typename] = cls
+
+        except IndexError:
+            # This is the base class, don't register it (bases was empty)
+            pass
+
+        except AttributeError:
+            # The class doesn't have a 'typename' attribute, don't register it
+            pass
+
+        return cls
+
+
+class Package(metaclass=MetaRegistry):
+    registered_types = {}
+
+    def __init__(self, id, metadata):
+        self.id = id
+        self._metadata = metadata
+
+    def __eq__(self, other):
+        return self.id == other.id and self._metadata == other._metadata
+
+    def __getattr__(self, name):
+        try:
+            return self._metadata[name]
+
+        except KeyError:
+            raise AttributeError(name)
+
+    @property
+    def version(self):
+        # 0 has the advantage of always being "smaller" than any other version
+        return self._metadata.get('version', '0')
+
+    def install(self, download_path, install_dir):
+        raise NotImplementedError('Subclasses must implement this method')
+
+    def remove(self, install_dir):
+        raise NotImplementedError('Subclasses must implement this method')
+
+
+class ZippedZim(Package):
+    typename = 'zipped-zim'
+
+    def install(self, download_path, install_dir):
+        if not zipfile.is_zipfile(download_path):
+            os.unlink(download_path)
+            raise InvalidFile('{} is not a zip file'.format(download_path))
+
+        with zipfile.ZipFile(download_path, "r") as z:
+            names = z.namelist()
+            names = filter(lambda n: n.startswith('data/'), names)
+            z.extractall(install_dir, members=names)
```

Comment:
Real life zips will take a lot to unzip (47Go for English Wikipedia or such), particularly in Koombook. I wonder if we don't need a progress bar here too?</autopromo> :)